design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,201 @@
+* Read from file to allow reproducible insertions
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+```sql
+CREATE TABLE table_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+The `WITH` clause will now accept two additional property_names: `PARTITIONS`/`REPLICAS` If the 
+topic *does not exist* and these values are specified, this command will create a topic with a the
+set number of partitions and replicas before executing the command. If the topic exists and has
+a different number of partitions/replicas, this command will fail.
+
+This functionality will be flagged under a configuration `ksql.create.source.allow.empty.topic`,
+which will default to `true` (new behavior is the default).
+
+## Design
+
+### API Design
+
+The above APIs make various trade-offs, outlined below:
+
+* The proposal to overload `INSERT INTO` is to remain as compatible with SQL standard as possible.
+The parser should be lookahead and determine whether or not `VALUES` is present in order to leverage
+that data producing APIs. An alternative would be to use the word `PRODUCE`, which also makes it
+clear that it uses low-level Kafka producers.
+* Allowing `CREATE TABLE/STREAM` to specify a kafka topic that does not yet exist makes it easy to
+use the new APIs without any other system. There are three alternatives: 
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. In
+headless mode, the data will be produced on every machine that executes the SQL file if it contains
+an `INSERT INTO ... VALUES` statement. It is important to support these operations even in headless
+mode to build automated test pipelines.
```

Comment:
Can you elaborate on this? I'm having trouble understanding why it's preferable to build automated test pipelines via headless mode.

Reply:
This is more for the "system test" use case. Imagine that I have KSQL running in production with a given SQL file and I want to do some automated regression tests on a "golden" data set. I could just maintain that golden data set as a SQL file of insert statements, prepend it to my production SQL file, spin up a test cluster, and verify the output.

Reply:
This feature could, though, be a follow-up if I find that it adds a significant chunk of work.

Reply:
Hm, I sort of see the value in that, but it still seems strange to me since if you were to write a "system test" that did that, you'd still need a different tool (not KSQL running headless mode) to verify your expected outputs, so the marginal benefit of moving the production of test data into KSQL is low. 

I don't mean to pick at this excessively -- I just think it's odd for headless mode to support a statement that doesn't get distributed in interactive mode. You'd only ever put an `INSERT INTO ... VALUES` statement into a queries file if you only planned to run that queries file on a single server (otherwise you'd be inserting duplicate values), which you'd typically only do for testing, for which is feels like headless mode isn't the right vehicle.

Reply:
That makes sense. I will punt the headless mode question to another time and for the extent of this KLIP just stick to interactive. When we have a real story around system tests maybe I will revisit this.

Reply:
Personally, I see no reason to restrict headless to not support this new feature.  People will find weird and wonderful ways of using it, I'm sure.

If running a multi-node KSQL headless cluster, it would be easy enough to have only one node with the insert commands in its script...

Reply:
FYI https://github.com/confluentinc/ksql/pull/2693#discussion_r276322603 - if it was more work to restrict it, I'd probably just let it be. But since it's more work to support it, I'll (similarly) let it be.